### PR TITLE
Apip-609 Host header replacement

### PR DESCRIPTION
**Pull request by medalliaerlich**
Added filter to allow replacement of the Request's header with the value of the "host" element in the "Forwarded" header.
This is **not** configurable because we expect all the requests to be sent with the actual host value and not the service name that L5D uses for lookup.

Test strategy:

- Local/functional: Local ZK instance with Local L5D, manual requests to mockbin.org/request with or without the Forwarded header with or without the Host element, and checking the mockbin response.
- Unit tests:
-- Doesn't fail if neither host or forwarded headers are available
-- If Forwarded is not present host should not change
-- If Forwarded is present but without host element, request's host should not change
-- Host is changed if Forwarded header comes with host element
-- Host is changed if Forwarded header comes with host element and others
-- Host is changed if Forwarded header comes with host element and others, with spaces and empty elements
-- Host is changed if Forwarded header comes with host element and others, is not case sensitive

### CODE REVIEW

**Comment by medalliaerlich** [comment]
@fiore-adrian and @erdody, please review this, I don't have permissions to assign code reviewers.

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    svc(req)
+  }
+
+  private def getForwardedHost(headerMap: HeaderMap): Option[String] = {
+    headerMap.get(forwardedHeader)
+      .flatMap { f =>
+        f.split(";").toStream
+          .map(_.trim)
+          .find(x => x.toLowerCase().startsWith("host="))
```
Comment (erdody) [inline router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: StringUtils.startsWithIgnoreCase ?

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import java.net.{Inet4Address, Inet6Address, InetSocketAddress, SocketAddress, URI, URISyntaxException}
+import scala.collection.mutable
+import scala.util.Random
+import com.twitter.logging.Logger
+
+/**
+ * Applies the Host value in the [Forwarded](https://tools.ietf.org/html/rfc7239) header to the Request's Host header.
+ * So the Request is sent with the actual host value.
+ */
+class ApplyHostForwardedHeader() extends SimpleFilter[Request, Response] {
```
Comment (erdody) [inline router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Can we make this an object? (look at ViaHeaderAppenderFilter)

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  private val log = Logger(getClass)
+
+  val forwardedHeader = "Forwarded"
```
Comment (erdody) [inline router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: the convention is to start constants with uppercase.

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  class HostHeaderReplacementProxy(
+    underlying: ServiceFactory[Request, Response]
+  ) extends ServiceFactoryProxy(underlying) {
+
+    override def apply(conn: ClientConnection): Future[Service[Request, Response]] = {
+      val filter = new ApplyHostForwardedHeader()
+      self.apply(conn).map(filter.andThen(_))
+    }
+  }
+  val module: Stackable[ServiceFactory[Request, Response]] =
```
Comment (erdody) [inline router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Not sure if we need all this. 
Can we do something like what ViaHeaderAppenderFilter does? (the module object is just 4 lines)

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  val forwardedHeader = "Forwarded"
+  def apply(req: Request, svc: Service[Request, Response]): Future[Response] = {
+
+    replaceHostWithForwardedHostIfExists(req)
+
+    svc(req)
+  }
+
+  private def getForwardedHost(headerMap: HeaderMap): Option[String] = {
```
Comment (erdody) [inline router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Should we support comma separated fields too?

From the RFC:

> Examples:
> 
>        Forwarded: for="_gazonk"
>        Forwarded: For="[2001:db8:cafe::17]:4711"
>        Forwarded: for=192.0.2.60;proto=http;by=203.0.113.43
>        Forwarded: for=192.0.2.43, for=198.51.100.17
> 
>    Note that as ":" and "[]" are not valid characters in "token", IPv6
>    addresses are written as "quoted-string".
> 
>    A proxy server that wants to add a new "Forwarded" header field value
>    can either append it to the last existing "Forwarded" header field
>    after a comma separator or add a new field at the end of the header
>    block.

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    headerMap.get(forwardedHeader)
+      .flatMap { f =>
+        f.split(";").toStream
+          .map(_.trim)
```
Comment (erdody) [inline router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Also, apparently the fields can be quoted.

**Review by erdody — CHANGES_REQUESTED** [review]

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala (hunk 2 above)
Comment (medalliaerlich) [reply router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Ok, done

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala (hunk 3 above)
Comment (medalliaerlich) [reply router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Ok, done

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala (hunk 5 above)
Comment (medalliaerlich) [reply router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Done

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala (hunk 1 above)
Comment (medalliaerlich) [reply router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Done, commons-lang3 was not a dependency so I used StringUtils from commons-lang which was.
I'm not sure of the impact of adding a new dependency right now so I didn't add it, is that OK with you?

**Comment by medalliaerlich** [comment]
@erdody, I've made all the requested changes.

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    private def getForwardedHost(headerMap: HeaderMap): Option[String] = {
+      headerMap.get(ForwardedHeader)
+        .flatMap { f =>
+          f.split(";|,").toStream
+            .map(_.trim)
+            .find(x => StringUtils.startsWithIgnoreCase(x, "host="))
+        }
+        .map(fHost => fHost.toLowerCase().replace("host=", "").replaceAll("\"|'", ""))
```
Comment (erdody) [inline router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: What do you think about CharMatcher.is().trimFrom, to only remove leading / trailing quotes?

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+      svc(req)
+    }
+
+    private def getForwardedHost(headerMap: HeaderMap): Option[String] = {
+      headerMap.get(ForwardedHeader)
+        .flatMap { f =>
+          f.split(";|,").toStream
+            .map(_.trim)
+            .find(x => StringUtils.startsWithIgnoreCase(x, "host="))
```
Comment (erdody) [inline router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Let's add a test with multiple "host", to verify/document that we pick the first one.

Comment (erdody) [inline router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Extract constant for "host="?

**Review by erdody — CHANGES_REQUESTED** [review]

File: router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala (hunk 1 above)
Comment (erdody) [reply router/http/src/main/scala/io/buoyant/router/http/ApplyHostForwardedHeader.scala]: Yep, that's fine.

**Comment by medalliaerlich** [comment]
Look now @erdody

**Review by erdody — APPROVED** [review]

**Pull request merged**